### ISSUE TITLE
Panels don't draw when using VSCode's debug mode under Windows (other OS's work fine, as does non-debug mode)

### ISSUE THREAD

**Issue by ddkwork** [richardwilkes/unison#51]
On windows, the demo in goland debug mode renders the layout but doesn't show up on the main window.
How to reproduce:
Goland opens unison demo, clicks debug, sees no content in the panel inside the window, the
After careful troubleshooting, the following commit id that causes this bug:

42c26e1249b14c404a071392a2a6a51525d7b4d8 ok ，last working version

d845b56964c1c227fde3bd819d5656818d43645a bug start

**Comment by richardwilkes** [comment]
I almost never use Windows, so haven't seen this when debugging. When I get some spare time, I'll take a look, though. Given my current workload, that may not be soon.

**Comment by ddkwork** [comment]
thx



---Original---
From: "Richard ***@***.***&gt;
Date: Thu, Apr 18, 2024 09:00 AM
To: ***@***.***&gt;;
Cc: ***@***.******@***.***&gt;;
Subject: Re: [richardwilkes/unison] A certain commit caused a panel renderingerror (Issue #51)




 
I almost never use Windows, so haven't seen this when debugging. When I get some spare time, I'll take a look, though. Given my current workload, that may not be soon.

**Comment by ddkwork** [comment]
I feel that we should start by checking the Window's root panel, because the process of rendering management should be like this: glfw creates the window and fills it into the Window structure, and there is a root panel inside the Window structure that manages all the layouts of the panel, and at the same time all the layouts' inputcallbacks are passed back to theAt the same time, the inputcallbacks of all the layouts are passed back to the inputcallback of the Window structure where the root panel is located, so the problem should be caused by some mysterious judgment that interrupts the root panel from submitting rendering data to the glfw manager, so I should start to study the interaction process between the root panel and glfw. The puzzling thing is that non-debug mode is normal, I seem to recall that skia's internal package has a switch for whether to enable debug mode, not sure as this is a code snippet I saw a long time ago. What do you think about a variable, or flag, that causes a certain judgment to interrupt the root panel in debug mode from submitting data normally, I think that's it?

**Comment by richardwilkes** [comment]
I've now had a chance to verify that debugging on Windows with any version beyond 42c26e1 causes all panels to be blank. Running without the debugger works fine, as does using macOS or Linux, whether in debug mode or not.

Since I don't use Windows for debugging, this doesn't affect me. I'd like to find a solution, but nothing obvious in that commit was changed that should affect rendering like this and my time is limited, so this is likely to remain open until someone else who really needs to be able to debug on Windows can spend the time to determine what the actual issue  is.

**Comment by ddkwork** [comment]
> I've now had a chance to verify that debugging on Windows with any version beyond 42c26e1 causes all panels to be blank. Running without the debugger works fine, as does using macOS or Linux, whether in debug mode or not.
> 
> Since I don't use Windows for debugging, this doesn't affect me. I'd like to find a solution, but nothing obvious in that commit was changed that should affect rendering like this and my time is limited, so this is likely to remain open until someone else who really needs to be able to debug on Windows can spend the time to determine what the actual issue  is.

I spent a few days checking the commits related to root panel and Window draw but no useful error messages appeared, I thought it could only be related to the compilation parameters of skia, unfortunately it's quite difficult to access github in China, and it's a bit overwhelming to clone a few gb's worth of skia repositories.

**Comment by ddkwork** [comment]
I spent a few days checking the commits related to root panel and Window draw but no useful error messages appeared, I thought it could only be related to the compilation parameters of skia, unfortunately it's quite difficult to access github in China, and it's a bit overwhelming to clone a few gb's worth of skia repositories.

**Comment by richardwilkes** [comment]
My first thought was also that skia had changed in some manner... however, there was no update of the skia library between the version that works and the one that doesn't.

**Comment by ddkwork** [comment]
I think we can try this: compile skia and compile the demo page that comes with the skia lib, the c one, or the cpp one, and then we can debug the c or cpp demo, which can be used to troubleshoot the bugs from the source.

**Comment by ddkwork** [comment]
Based on the experience of using it these days, it is very likely that the parent of a panel is set wrongly. Because I had a similar problem when I was designing this debug interface using unison.
https://github.com/HyperDbg/gui
What's interesting is that you can manipulate the panel even though you don't see it being rendered, assuming you know where each widget is located.

**Comment by richardwilkes** [comment]
Not likely, if you can actually interact with it despite it not drawing. The layout would not place things in the correct place if the parent was incorrect, nor would any events reach it.

That also doesn't explain why it works fine without the debugger attached.

**Comment by ddkwork** [comment]
👍👍👍我想你是对的



---Original---
From: "Richard ***@***.***&gt;
Date: Fri, Jun 28, 2024 23:18 PM
To: ***@***.***&gt;;
Cc: ***@***.******@***.***&gt;;
Subject: Re: [richardwilkes/unison] Panels don't draw when using VSCode'sdebug mode under Windows (other OS's work fine, as does non-debug mode) (Issue#51)




 
Not likely, if you can actually interact with it despite it not drawing. The layout would not place things in the correct place if the parent was incorrect, nor would any events reach it.
 
That also doesn't explain why it works fine without the debugger attached.

**Comment by ddkwork** [comment]
This catches some errors, but nothing to go on, do you have any instructions? thx

```go
	wnd.MouseEnterCallback = func(where unison.Point, mod unison.Modifiers) bool {
		wnd.SetFocus(wnd.Content())
		wnd.Focus()
		//wnd.FocusNext()
		wnd.SetContent(unison.NewButton())
		wnd.Wnd().SwapBuffers()
		wnd.Wnd().Show()
		return true
	}

INF | 2024-07-15 | 02:23:00.797 | PlatformError: WGL: Failed to make context current: The handle is invalid. 
INF | 2024-07-15 | 02:23:00.985 | PlatformError: WGL: Failed to make context current: The requested transformation operation is not supported. 
INF | 2024-07-15 | 02:23:01.091 | PlatformError: WGL: Failed to make context current: The requested transformation operation is not supported. 
INF | 2024-07-15 | 02:23:01.506 | PlatformError: WGL: Failed to make context current: The handle is invalid. 
INF | 2024-07-15 | 02:23:01.664 | PlatformError: WGL: Failed to make context current: The requested transformation operation is not supported. 
INF | 2024-07-15 | 02:23:01.881 | PlatformError: WGL: Failed to make context current: The handle is invalid. 
INF | 2024-07-15 | 02:23:01.956 | PlatformError: WGL: Failed to make context current: The requested transformation operation is not supported. 
INF | 2024-07-15 | 02:23:02.030 | PlatformError: WGL: Failed to make context current: The handle is invalid. 
INF | 2024-07-15 | 02:23:02.047 | PlatformError: WGL: Failed to make context current: The requested transformation operation is not supported. 
INF | 2024-07-15 | 02:23:02.050 | PlatformError: WGL: Failed to make context current: The requested transformation operation is not supported. 
INF | 2024-07-15 | 02:23:03.434 | PlatformError: WGL: Failed to make context current: The handle is invalid. 
INF | 2024-07-15 | 02:23:03.707 | PlatformError: WGL: Failed to make context current: The handle is invalid. 
INF | 2024-07-15 | 02:23:03.756 | PlatformError: WGL: Failed to make context current: The handle is invalid. 
INF | 2024-07-15 | 02:23:03.838 | PlatformError: WGL: Failed to make context current: The requested transformation operation is not supported. 
INF | 2024-07-15 | 02:23:04.146 | PlatformError: WGL: Failed to make context current: The handle is invalid. 
INF | 2024-07-15 | 02:23:04.160 | PlatformError: WGL: Failed to make context current: The requested transformation operation is not supported. 
INF | 2024-07-15 | 02:23:04.264 | PlatformError: WGL: Failed to make context current: The requested transformation operation is not supported. 
```

**Comment by richardwilkes** [comment]
This is an invalid use of SwapBuffers. You will get random results.

**Comment by ddkwork** [comment]
oh



---Original---
From: "Richard ***@***.***&gt;
Date: Mon, Jul 15, 2024 02:42 AM
To: ***@***.***&gt;;
Cc: ***@***.******@***.***&gt;;
Subject: Re: [richardwilkes/unison] Panels don't draw when using VSCode'sdebug mode under Windows (other OS's work fine, as does non-debug mode) (Issue#51)




 
This is an invalid use of SwapBuffers. You will get random results.

**Comment by ddkwork** [comment]
```go
	wnd.MouseEnterCallback = func(where unison.Point, mod unison.Modifiers) bool {
		runtime.LockOSThread()
		//wnd.SetFocus(wnd.Content())
		//wnd.Focus()
		////wnd.FocusNext()
		//wnd.Wnd().Focus()
		wnd.SetContent(unison.NewButton())
		//wnd.Wnd().SwapBuffers()
		//wnd.Wnd().Show()
		return true
	}


```

No, an error can occur without a swap buffer. Can this be an entry point?

**Comment by richardwilkes** [comment]
The use of LockOSThread is also not appropriate here and will cause problems.

Why do you think mouse enter events are relevant?

**Comment by ddkwork** [comment]
Locking threads commented out can also be captured, as there is no better tangent event to immediately trigger setting the context of the window

**Comment by richardwilkes** [comment]
You're still not making any sense. What are you trying to accomplish? The OpenGL context is set appropriately at draw time and that's the only time it needs to be dealt with. Unless you've gone out of your way to break the way drawing is done, this is all automatic in unison.

**Comment by ddkwork** [comment]
![910d3285-f855-4e0d-ace5-8f79008a9b7c.png](https://github.com/user-attachments/assets/93ac7db2-a4f2-4994-b226-6fd170361253)

**Comment by richardwilkes** [comment]
Yes, I get that there is some interaction on Windows with the debugger that is causing the OpenGL canvas to not draw -- that's the subject of this issue. However, mouse events are not related to this in any way that I can determine, so why are you messing with that aspect?

**Comment by ddkwork** [comment]
Didn't I explain that before? Other than that there is no immediate way to trigger the bug, unless you have one, please tell me how to force the bug to be triggered
![444ea5a4-2f88-4bdb-8367-dc6861c72b7a.png](https://github.com/user-attachments/assets/dd2e153d-f15d-4f78-bdda-0a492621ab86)

**Comment by richardwilkes** [comment]
For me -- on Windows only -- if I run the demo with the debugger, it just never draws anything in the window. That's the bug. No mucking about with the mouse events matters. Why this happens, I don't know. It appears that the Windows version of the debugger is somehow messing around with the OpenGL context, which it should not be doing -- and doesn't on all other platforms.

Anyway, this is not a problem for me, as I almost never use Windows for development, as it is an extremely poor platform for such activities. I only ever use it when debugging a problem, and I can get by with printf-style debugging in those cases where I absolutely must use Windows to debug a problem. Because of this, I've not invested much time or energy into trying to determine why Windows is doing the wrong thing.

**Comment by ddkwork** [comment]
That's okay, I'll try to find some time to see if I can figure it out. Since I'm only using Windows at the moment, Linux is only used occasionally. Interestingly, the screenshot just now is the same behavior in debug mode for that commit id I specified, and it didn't throw an error.

Maybe I should just wrap the widget back to the old version and upgrade the new version when it's resolved. Because this kind of problem is unlikely to be solved anytime soon.

If you are too busy, you can close this issue.

**Comment by richardwilkes** [comment]
I'll leave it open, as it is a real issue.

**Comment by ddkwork** [comment]
After debugging glfw and finding that the corrupted context or handle comes from the api parameter before gl handles the drawing and it comes from skia, it looks like the bug comes from skia and I need to test the master branch of skia

**Comment by ddkwork** [comment]
Hi, after testing the main branch skia still has this problem, then m128 should be the same, but the old version I guess is no longer compatible with the c api, can you give me a version between m118-m127 that probably might work? If switching multiple versions doesn't work, then you can narrow down the troubleshooting. thanks!!!

**Comment by richardwilkes** [comment]
As I noted at the very start of this issue, there was no skia change between a working version and a non-working version, so it is unlikely to be in that area. It might be in the version of glfw that was used. I've not checked that.

**Comment by ddkwork** [comment]
I solved the problem by downgrading the unison version, so close this topic

**Comment by ddkwork** [comment]
some test code
```cpp
#include "GLFW/glfw3.h"
#define SK_GANESH
#define SK_GL
#include "include/gpu/GrBackendSurface.h"
#include "include/gpu/GrDirectContext.h"
#include "include/gpu/gl/GrGLInterface.h"
#include "include/gpu/gl/GrGLAssembleInterface.h"
#include "include/gpu/ganesh/SkSurfaceGanesh.h"
#include "include/gpu/ganesh/gl/GrGLBackendSurface.h"
#include "include/gpu/ganesh/gl/GrGLDirectContext.h"
#include "include/core/SkCanvas.h"
#include "include/core/SkColorSpace.h"
#include "include/core/SkSurface.h"
#include "include/core/SkGraphics.h"

#include <stdio.h>
#include <stdlib.h>

//uncomment the two lines below to enable correct color spaces
//#define GL_FRAMEBUFFER_SRGB 0x8DB9
//#define GL_SRGB8_ALPHA8 0x8C43

GrDirectContext *sContext = nullptr;
SkSurface *sSurface = nullptr;

void error_callback(int error, const char *description) {
    fputs(description, stderr);
}

void key_callback(GLFWwindow *window, int key, int scancode, int action, int mods) {
    if (key == GLFW_KEY_ESCAPE && action == GLFW_PRESS)
        glfwSetWindowShouldClose(window, GL_TRUE);
}

void init_skia(int w, int h) {
    auto interface = GrGLMakeNativeInterface();
    if (interface == nullptr) {
        //backup plan. see https://gist.github.com/ad8e/dd150b775ae6aa4d5cf1a092e4713add?permalink_comment_id=4680136#gistcomment-4680136
        interface = GrGLMakeAssembledInterface(nullptr, (GrGLGetProc) &[](void *, const char *p) -> void *{
            return (void *) glfwGetProcAddress(p);
        });
    }
    sContext = GrDirectContexts::MakeGL(interface).release();

    GrGLFramebufferInfo framebufferInfo;
    framebufferInfo.fFBOID = 0; // assume default framebuffer
    // We are always using OpenGL and we use RGBA8 internal format for both RGBA and BGRA configs in OpenGL.
    //(replace line below with this one to enable correct color spaces) framebufferInfo.fFormat = GL_SRGB8_ALPHA8;
    framebufferInfo.fFormat = GL_RGBA8;

    SkColorType colorType = kRGBA_8888_SkColorType;
    GrBackendRenderTarget backendRenderTarget = GrBackendRenderTargets::MakeGL(
        w,
        h,
        1, // sample count
        8, // stencil bits
        framebufferInfo);

    //(replace line below with this one to enable correct color spaces) sSurface = SkSurfaces::WrapBackendRenderTarget(sContext, backendRenderTarget, kBottomLeft_GrSurfaceOrigin, colorType, SkColorSpace::MakeSRGB(), nullptr).release();
    sSurface = SkSurfaces::WrapBackendRenderTarget(sContext, backendRenderTarget, kBottomLeft_GrSurfaceOrigin,
                                                   colorType, nullptr, nullptr).release();
    if (sSurface == nullptr) abort();
}

void cleanup_skia() {
    delete sSurface;
    delete sContext;
}

const int kWidth = 960;
const int kHeight = 640;


int main(void) {
    GLFWwindow *window;
    glfwSetErrorCallback(error_callback);
    if (!glfwInit()) {
        exit(EXIT_FAILURE);
    }

    glfwWindowHint(GLFW_CONTEXT_VERSION_MAJOR, 3);
   glfwWindowHint(GLFW_CONTEXT_VERSION_MINOR, 2);
    glfwWindowHint(GLFW_OPENGL_FORWARD_COMPAT, GL_TRUE);
    glfwWindowHint(GLFW_OPENGL_PROFILE, GLFW_OPENGL_CORE_PROFILE);
    //(uncomment to enable correct color spaces) glfwWindowHint(GLFW_SRGB_CAPABLE, GL_TRUE);
    glfwWindowHint(GLFW_STENCIL_BITS, 0);
    //glfwWindowHint(GLFW_ALPHA_BITS, 0);
    glfwWindowHint(GLFW_DEPTH_BITS, 0);

    window = glfwCreateWindow(kWidth, kHeight, "Simple example", NULL, NULL);
    if (!window) {
        glfwTerminate();
        exit(EXIT_FAILURE);
    }
    glfwMakeContextCurrent(window);
    //(uncomment to enable correct color spaces) glEnable(GL_FRAMEBUFFER_SRGB);

    init_skia(kWidth, kHeight);

    glfwSwapInterval(1);
    glfwSetKeyCallback(window, key_callback);
    // glfwSetMouseButtonCallback()

    // Draw to the surface via its SkCanvas.
    SkCanvas *canvas = sSurface->getCanvas(); // We don't manage this pointer's lifetime.

    while (!glfwWindowShouldClose(window)) {
        glfwWaitEvents();
        SkPaint paint;
        paint.setColor(SK_ColorWHITE);
        canvas->drawPaint(paint);
        paint.setColor(SK_ColorBLUE);
        canvas->drawRect({100, 200, 300, 500}, paint);
        // sContext->flush();
        sContext->flushAndSubmit();

        glfwSwapBuffers(window);
    }

    cleanup_skia();

    glfwDestroyWindow(window);
    glfwTerminate();
    exit(EXIT_SUCCESS);
}

```

**Comment by ddkwork** [comment]
Tested most of the versions of the go glfw source repository that I have modified, skia 110 129 128 127 and many more. Even upgrading the unison version of skia to m128 129, which is the version that the commit id points to, which functions normally, does not reproduce the bug in the above cpp demo, and multiple versions of glfw work fine. Ultimately had to downgrade, no choice.

**Comment by ddkwork** [comment]
I'll upgrade the current skia version to the main branch to determine if skia is buggy again after I implement more fixes. then turn on skia's gn gen parameter angle related to executing drawing api's for mobile platforms, as well as gradually replacing glfw with eb's backend, which will invoke both glfw and mobile backend support.

This process is too long. Once we bind the mobile rendering backend to the mobile drawing api, we can write a java entry wrapper function to encapsulate the mobile entry code, and a cmd tool to package it into an apk.

I now feel that you were right before, this is too complicated to be worth the hassle of doing this.

**Comment by richardwilkes** [comment]
Unfortunately, the last time I looked at vulkan (admittedly, a couple of years ago), it had a fairly large number of problems and did not work well on all platforms.

**Comment by ddkwork** [comment]
> Unfortunately, the last time I looked at vulkan (admittedly, a couple of years ago), it had a fairly large number of problems and did not work well on all platforms.

There does need to be more success stories, and besides, VMware doesn't seem to support vulkan.

**Comment by ddkwork** [comment]
> Unfortunately, the last time I looked at vulkan (admittedly, a couple of years ago), it had a fairly large number of problems and did not work well on all platforms.

I don't know if my logic is correct :)

Let's change the desktop platform's vulkan rendering to the current opengl rendering, https://github.com/cogentcore/core/blob/main/system%2Fdriver%2Fdesktop%2Fapp.go#L136-L139

Since the drawing for the widgets comes from skia, we just need to bind skia's mobile drawing api and call the vulkan backend under the android package, with skia as the frontend.

This way, the drawing api for all platforms comes from skia and should be consistent. While we removed the 3d drawing functions from that repository, we preserved the current backend implementation, which should be enough to keep the ported code stable on desktop platforms. As for the stability of android under vulkan, you can take your time to study it.

**Comment by ddkwork** [comment]
Also, there's quite a lot of stuff that needs to be changed to copy its code over and change his threaded event system to unison's non-threaded callback mode. I think we'll just use his backend related code and leave the rest.

What I'm a bit unsure about though is:desktop uses glfw, android uses vulkan, and all platform drawing uses skia.
Is this combination really logical? Looking at his implementation, we need to do the following steps.

1. copy the code from the system package, nothing else.
2. remove the desktop vulkan rendering logic and replace it with gl and skia interactions
3. bind skia's android drawing api.
4. make apk build cmd tool

**Comment by ddkwork** [comment]
You're right, after a deep investigation of vulkan, I ended up switching back to gles for mobile and web backend support. Also, I agree that subpackaging adds complexity as your readme states. When I introduce web and mobile backend code, a lot of go files are added to reach conditional compilation, which makes the directory tree longer. Just now I implemented widget and layout subcontracting management, which is really inconvenient, I'll have to change it back.
